runtime/runtime/src/congestion_control.rs

```diff
@@ -91,8 +103,22 @@ impl<'a> ReceiptSink<'a> {
                         congestion.congestion_info,
                         congestion.missed_chunks_count,
                     );
-
-                    (shard_id, other_congestion_control.outgoing_limit(apply_state.shard_id))
+                    let mut gas_limit =
+                        other_congestion_control.outgoing_gas_limit(apply_state.shard_id);
+                    if shard_id == apply_state.shard_id {
+                        // No gas limits on receipts that stay on the same shard. Backpressure
+                        // wouldn't help, the receipt takes the same memory if buffered or
+                        // in the delayed receipts queue.
+                        gas_limit = Gas::MAX;
+                    }
+
+                    let size_limit = if special_shard == Some(shard_id) {
+                        apply_state.config.witness_config.outgoing_receipts_big_size_limit
+                    } else {
+                        apply_state.config.witness_config.outgoing_receipts_usual_size_limit
+                    };
```

Comment:
How exactly should that work?
AFAIU when all shards are fully congested one shard is allowed to process receipts to avoid deadlocks.
But I can't set the special shard to this allowed shard for every sender shard, because then multiple senders would be allowed to send large receipts to this one shard, and `source_receipt_proofs` for the allowed shards could explode.
I don't fully understand how it works in congestion control, so maybe I'm missing something.

Reply:
Thanks, I see that I misunderstood how that works. I'll read into the code and try to merge them.

Reply:
One small difference is that `special_shard` can be equal to `own_shard`, but `allowed_shard` is supposed to be different from `own_shard`. I guess I'll just allow `allowed_shard` to be the same as `own_shard`. It might lower the throughput in super-high congestion scenarios, but it's at most 1/num_shards of a difference.